crates/llm-local/src/lib.rs

```diff
@@ -1,35 +1,60 @@
 mod bert;
+mod llama;
+mod token_output_stream;
 
 use anyhow::Context;
 use bert::{BertModel, Config};
-use candle::DType;
+use candle::{safetensors::load_buffer, DType};
 use candle_nn::VarBuilder;
-use llm::{
-    InferenceFeedback, InferenceParameters, InferenceResponse, InferenceSessionConfig, Model,
-    ModelArchitecture, ModelKVMemoryType, ModelParameters,
-};
-use rand::SeedableRng;
 use spin_common::ui::quoted_path;
+use spin_core::async_trait;
 use spin_world::v2::llm::{self as wasi_llm};
 use std::{
-    collections::hash_map::Entry,
-    collections::HashMap,
-    convert::Infallible,
+    collections::{hash_map::Entry, HashMap},
     path::{Path, PathBuf},
-    sync::{Arc, Mutex},
+    str::FromStr,
+    sync::Arc,
 };
 use tokenizers::PaddingParams;
 
 const MODEL_ALL_MINILM_L6_V2: &str = "all-minilm-l6-v2";
+type ModelName = String;
 
 #[derive(Clone)]
 pub struct LocalLlmEngine {
     registry: PathBuf,
-    use_gpu: bool,
-    inferencing_models: HashMap<(String, bool), Arc<dyn llm::Model>>,
+    inferencing_models: HashMap<ModelName, Arc<dyn CachedInferencingModel>>,
     embeddings_models: HashMap<String, Arc<(tokenizers::Tokenizer, BertModel)>>,
 }
 
+#[derive(Debug)]
+enum InferencingModelArch {
+    Llama,
+}
+
+impl FromStr for InferencingModelArch {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "llama" => Ok(InferencingModelArch::Llama),
+            _ => Err(()),
+        }
+    }
+}
+
+/// `CachedInferencingModel` implies that the model is prepared and cached after
+/// loading, allowing faster future requests by avoiding repeated file reads
+/// and decoding. This trait does not specify anything about if the results are cached.
```

Comment:
Formatting nit - typically you want a single line descriptor, a blank line, and then further information:

```suggestion
/// A model that is prepared and cached after loading.
///
/// This trait does not specify anything about if the results are cached.
```